apis/v1alpha1/backendpolicy_types.go

```diff
@@ -0,0 +1,220 @@
+	// following to specify a service:
+	//
+	// fooRef:
+	//   resource: services
+	//   name: myservice
+	//
+	// Otherwise, if the core API group is not desired, specify the desired
+	// group:
+	//
+	// fooRef:
+	//   group: acme.io
+	//   resource: foos
+	//   name: myfoo
+	//
+	// +optional
+	// +kubebuilder:default=core
+	Group string `json:"group" protobuf:"bytes,1,opt,name=group"`
+
+	// Resource is the API resource name of the referent. Omitting the value
+	// or specifying the empty string indicates the services resource. For example,
+	// use the following to specify a services resource:
+	//
+	// fooRef:
+	//   name: myservice
+	//
+	// Otherwise, if the services resource is not desired, specify the desired
+	// group:
+	//
+	// fooRef:
+	//   group: acme.io
+	//   resource: foos
+	//   name: myfoo
+	//
+	// +optional
+	// +kubebuilder:default=services
+	Resource string `json:"resource" protobuf:"bytes,2,opt,name=resource"`
+
+	// Name is the name of the referent.
+	//
+	// +kubebuilder:validation:Required
+	Name string `json:"name" protobuf:"bytes,3,opt,name=name"`
+
+	// Port is the port of the referent. If unspecified, this policy applies to
+	// all ports on the backend.
+	// +optional
+	Port *int32 `json:"port,omitempty" protobuf:"varint,4,opt,name=port"`
+}
+
+// BackendTLSConfig describes TLS configuration for a backend.
+type BackendTLSConfig struct {
+	// ClientCertificateRef is a reference to a TLS client certificate that may
+	// be used to connect to these backends. If an entry in this list omits or
+	// specifies the empty string for both the group and the resource, the
+	// resource defaults to "secrets". An implementation may support other
+	// resources (for example, resource "mycertificates" in group
+	// "networking.acme.io").
+	//
+	// Support: Extended
+	//
+	// +optional
+	ClientCertificateRef *CertificateObjectReference `json:"clientCertificateRef,omitempty" protobuf:"bytes,1,rep,name=clientCertificateRef"`
+
+	// ClientCertificateKeyRef is a reference to a TLS client certificate key
+	// that may be used to connect to these backends. If an entry in this list
+	// omits or specifies the empty string for both the group and the resource,
+	// the resource defaults to "secrets". An implementation may support other
+	// resources (for example, resource "mycertificates" in group
+	// "networking.acme.io").
+	//
+	// Support: Extended
+	//
+	// +optional
+	ClientCertificateKeyRef *CertificateObjectReference `json:"clientCertificateKeyRef,omitempty" protobuf:"bytes,2,rep,name=clientCertificateKeyRef"`
+
+	// TrustedCertificateRef is a reference to a certificate for a trusted
+	// certificate authority. This may be useful for backends using self-signed
+	// certificates.
+	//
+	// Support: Extended
+	//
+	// +optional
+	TrustedCertificateRef *CertificateObjectReference `json:"trustedCertificateRef,omitempty" protobuf:"bytes,3,rep,name=trustedCertificateRef"`
```

Comment:
This should be an array as there can be multiple CA certs, especially during certificate rotation periods.

Is `CACertifcateRefs` a better name here (which indicates that the user needs to put in a reference to a CA certificate?

Since this is an extended feature, we should have a standardized way (key name, content formatting etc) of defining certificates using secrets or config maps for this feature.

We don't need to do this for client cert-key because we can directly rely on TLS secrets.

Reply:
I initially transitioned to `CACertificateRefs` but in the last meeting I think @jpeach mentioned that we could include a CA Bundle inside a TLS secret. That is a configuration currently [supported by nginx-ingress](https://kubernetes.github.io/ingress-nginx/examples/auth/client-certs/), though does not seem to be as broadly supported or used by other implementations. 

A potential benefit of having everything in the same resource is that there are less Secrets to maintain. A downside is that some use cases may involve reusing a CA bundle but not client certs, something that would be a bit more difficult with everything in a single Secret.

Allowing `ca.crt` to be specified at this level does not prevent us from adding support for a separate field to reference a CA bundle in the future. Given that, I think this simple approach is likely the best starting point.

Reply:
I don't agree here.

First, CA certificates are usually managed separately from client certificates. They have different life-cycles and clubbing them together often leads to unnecessary duplication of CA certificates.

Further, the doc states:

> 	// If a Secret is referenced, it must be of type "kubernetes.io/tls" and
	// contain tls.crt and tls.key data fields that contain the certificate and
	// private key to use for TLS. It may optionally include a certificate
	// authority bundle under a ca.crt data field



Is that even valid?
The `kubectl` document says `kubectl create secret tls tls-secret --cert=path/to/tls.cert --key=path/to/tls.key` is how you create a tls secret and it doesn't allow you to specify any additional field. You can edit a secret or manually (using stringData) create a secret of type `kubernetes.io/tls` that allows you to include `ca.crt`.

The [nginx document](https://kubernetes.github.io/ingress-nginx/examples/auth/client-certs/) asks you to create a generic secret instead of a TLS secret:
```
kubectl create secret generic ca-secret --from-file=tls.crt=server.crt --from-file=tls.key=server.key --from-file=ca.crt=ca.crt
```

Reply:
Just throwing some additional info our there, Istio supports the two options nginx-ingress supports (single secret with ca.crt or split secrets) and doesn't care if its `generic` or `kubernetes.io/tls` due to the issue mentioned ^ with `kubectl create`

Reply:
Split this back up into 2 refs, thanks for the feedback!